doc/developer/design/20220413_cluster_replica.md

```diff
@@ -0,0 +1,109 @@
+# Cluster replicas
+
+## Summary
+
+This document proposes an explicit SQL API for managing
+[cluster replicas](../platform/ux.md#cluster-replica).
+
+## Design
+
+### SQL syntax and semantics
+
+The `ALTER CLUSTER` statement will be removed.
+
+The `CREATE CLUSTER` and `CREATE CLUSTER REPLICA` statement will be evolved as
+follows:
+
+```sql
+create_cluster_stmt ::=
+  CREATE CLUSTER [IF NOT EXISTS] <name>
+    [<inline_replica> [, <inline_replica> ...]]
+
+create_cluster_replica_stmt ::=
+  CREATE CLUSTER REPLICA [IF NOT EXISTS] <name> OF <cluster>
+    [<replica_option> [, <replica_option> ...]]
+
+inline_replica ::= REPLICA <name> [<replica_option> [, <replica_option> ...]]
+
+replica_option ::=
+    | SIZE '<size>'
+    | AVAILABILITY ZONE '<az>'
+    | REMOTE (['address' [, 'address' ...]])
+```
+
+It will be valid to create a cluster with no replicas. Creating a replica will
+be the moment at which compute resources are provisioned.
+
+The `SIZE` and `REMOTE` options work as they presently do today on the `CREATE
+CLUSTER` command. Users will be able to specify only one or the other, not both.
+
+The `AVAILABILITY ZONE` option is new. The set of allowable availability zones
+will be determined by a new `--allow-availability-zone=AZ` command-line option.
+
+The `DROP CLUSTER REPLICA` statement will be added:
+
+```
+DROP CLUSTER REPLICA [IF EXISTS] <name>
+```
+
+It will have the usual semantics for a `DROP` operation.
+
+The `SHOW CLUSTER REPLICAS` statement
+
+```
+SHOW CLUSTER REPLICAS [FOR <cluster>] [{ LIKE 'pattern' | WHERE <expr> }]
+```
+
+will list the replicas of the provided cluster, or of the current cluster if no
+cluster is specified. The results will be optionally filtered by the provided
+`LIKE` pattern or `WHERE` expression, which work analogously to the same clauses
+in the `SHOW SCHEMA` statement.
+
+### System catalog changes
+
+A new `mz_cluster_replicas` table will be added with the following schema:
+
+Field             | Type      | Meaning
+------------------|-----------|--------
+cluster_id        | `bigint`  | The ID of the cluster.
+name              | `text`    | The name of the cluster replica.
+size              | `text`    | The size of the cluster replica. `NULL` for a remote cluster.
+availability_zone | `text`    | The availability zone of the cluster replica, if any. May be `NULL`.
+status            | `text`    | The status of the replica. Either `unhealthy` or `healthy`.
+
+### Controller changes
+
+The controller will need to inform the coordinator when a replica changes
+status, so that the new status can be reflected in the `mz_cluster_replicas`
+table. The easiest way to do this is likely via a new
+`ComputeResponse::ReplicaStatusChange` variant.
```

Comment:
Can do, but might it also make sense to throw the actual information back up to the adapter, in the form of maybe `mz_materialization_frontiers` for the replicas from which one can look at the skew and reach one's own conclusions? Said differently: if compute owns this and it is just a bit, will adapter (proxying users) continually wish that it could be a different bit?

Reply:
Yeah, probably! It wasn't immediately clear to me how to use `mz_materialization_frontiers` to derive that bit in the case that you only have one replica, though!

Reply:
Let's leave it in! I think Maybe the meaning of this response changes a lot though, and eventually maybe we conclude some of its functionality should be elsewhere.